artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/LargeServerMessageImpl.java

```diff
@@ -70,8 +75,26 @@ private static Message asLargeMessage(Message message, StorageManager storageMan
       LargeServerMessage lsm = storageManager.createLargeMessage(storageManager.generateID(), coreMessage);
       ActiveMQBuffer buffer = coreMessage.getReadOnlyBodyBuffer();
       final int readableBytes = buffer.readableBytes();
-      lsm.addBytes(buffer);
-      lsm.releaseResources(true, true);
+
+      ByteBuf buf  = PooledByteBufAllocator.DEFAULT.buffer(CHUNK_LM_SIZE);
+      try {
+         ActiveMQBuffer wrappedBuffer = new ChannelBufferWrapper(buf);
+
+         // We write in chunks to avoid allocating a full NativeBody sized as the message size
+         // which might lead the broker out of resources
+         while (buffer.readableBytes() > 0) {
+            wrappedBuffer.writerIndex(0);
+            wrappedBuffer.readerIndex(0);
+            int bytesToRead = Math.min(CHUNK_LM_SIZE, buffer.readableBytes());
+            buffer.readBytes(wrappedBuffer, 0, bytesToRead);
```

Comment:
Couldnt the wrappedBuffer just be playing with the indices/limits so that 'windows' of the original underlying "buffer" so they can be added directly to the LSM incrementally rather than copied out of it and then immediately copied back into the LSM?

Reply:
@gemmellr not really... It is a little more subtle than this.. and I will a comment with this detail


The fact that I'm using a native buffer for this chunk, it means that later down the line, this will be translated as:

FileChannel::public abstract int write(ByteBuffer src) throws IOException;


if I send this as a non native buffer, it would be translated to write([] bytes) which will lead the system to allocate a random buffer.


This is actually the main reason the system was failing. because of these Ghost Native Buffers that we have no control over them.


At least here I have full control and wouldn't have any issues.



(that's a worth point to add a comment)

Reply:
I am adding the following comment:

      // I'm creating a native buffer here
      // because FileChannelImpl (which is used by NIOSequentialFile) would create a Ghost Native Buffer
      // that we would have no control. that's usually stored in a ThreadLocal within the native layer.
      // to avoid that buffer be kept in memory holding resources we will allocate our own buffer here from the NettyPool.
      // ./soakTest/OWLeakTest was writting to validate this scenario here.
      ByteBuf buf  = PooledByteBufAllocator.DEFAULT.buffer(CHUNK_LM_SIZE);